### PR TITLE
[RLlib] Mask phantom bootstrap timesteps from GAE advantage standardization

### LINKED ISSUE #57989: [RLlib] Possible issue with advantage standardization in general advantage estimation connector
### What happened + What you expected to happen

In the GeneralAdvantageEstimation ConnectorV2, the advantages are standardized wrt the mean and stdev of the entire batch. However unless I'm mistaken this doesn't take into account the fact that some of the timesteps will be masked in the loss (due to recurrence, but also even if not because of previous connectors like AddOneTsToEpisodesAndTruncate). Shouldn't the code read something more like this?

```python
loss_mask = unpad_data_if_necessary(episode_lens, batch[module_id][Columns.LOSS_MASK])

module_advantages = module_value_targets - module_vf_preds

module_advantages = (module_advantages - module_advantages.mean(where=loss_mask)) / max(
    1e-4, module_advantages.std(where=loss_mask)
)
```

For what it's worth, I took a brief look at the old stack (I'm not super familiar with rllib but I think I was looking in the right place) and from what I can tell it was handled correctly then.

While we're on the topic of this connector, there are a couple of other points I don't fully understand:

1. The device is set to None outside the loop over modules but then is written over within each loop - however it's not actually used within the loop, rather it is used after the loop has exited. Is the intention to just have the (global) device be the device of whichever module happens to be last?
2. I am having some difficulty following the calculation in compute_value_targets since if I understand correctly the input could in general consist of multiple episodes concatenated together. So then why append a single trailing 0 on to flat_values?

### Versions / Dependencies

ray[rllib]==2.47.1

### Reproduction script

Apologies for the formatting, I cannot get everything to display as I intend but hopefully this is relatively easy to follow.

```python
from ray.rllib.env.single_agent_episode import SingleAgentEpisode
import gymnasium
import torch
from ray.rllib.core.rl_module.apis.value_function_api import ValueFunctionAPI
from ray.rllib.core.rl_module.torch.torch_rl_module import TorchRLModule
from ray.rllib.connectors.learner.add_one_ts_to_episodes_and_truncate import AddOneTsToEpisodesAndTruncate
from ray.rllib.connectors.common.add_observations_from_episodes_to_batch import AddObservationsFromEpisodesToBatch
from ray.rllib.connectors.learner.add_columns_from_episodes_to_train_batch import AddColumnsFromEpisodesToTrainBatch
from ray.rllib.connectors.common.add_time_dim_to_batch_and_zero_pad import AddTimeDimToBatchAndZeroPad
from ray.rllib.connectors.common.add_states_from_episodes_to_batch import AddStatesFromEpisodesToBatch
from ray.rllib.connectors.common.batch_individual_items import BatchIndividualItems
from ray.rllib.connectors.common.numpy_to_tensor import NumpyToTensor
from ray.rllib.connectors.learner.general_advantage_estimation import GeneralAdvantageEstimation
from ray.rllib.core.rl_module.multi_rl_module import MultiRLModule
from typing import Dict, Any, Optional
from ray.rllib.utils.typing import TensorType

# dummy rl module

class RLModule(TorchRLModule, ValueFunctionAPI):
    def setup(self):
        self.embedding_f = torch.nn.Linear(4, 5)
        self.actor_f = torch.nn.Linear(5, 2)
        self.critic_f = torch.nn.Linear(5, 1)
    def _forward_embeddings(self, batch):
        return self.embedding_f(batch["obs"])
    def _forward(self, batch):
        embeddings = self._forward_embeddings(batch)
        action_dist_inputs = self.actor_f(embeddings)
        return {"action_dist_inputs": action_dist_inputs, "embeddings": embeddings}
    def compute_values(
        self,
        batch: Dict[str, Any],
        embeddings: Optional[Any] = None,
    ) -> TensorType:
        if embeddings is None and "embeddings" in batch:
            embeddings = batch["embeddings"]
        elif embeddings is None:
            embeddings = self._forward_embeddings(batch=batch)
        value = self.critic_f(embeddings).squeeze()
        return value

env = gymnasium.make("CartPole-v0")
rl_module = RLModule(observation_space=env.observation_space, action_space=env.action_space, model_config={})

# get 2 random episodes

episodes = []
for _ in range(2):
    episode = SingleAgentEpisode()
    obs, info = env.reset()
    episode.add_env_reset(obs, info)
    done = False
    while not done:
        action = env.action_space.sample()
        obs, reward, terminated, truncated, info = env.step(action)
        episode.add_env_step(obs, action, reward, info, terminated=terminated, truncated=truncated)
        done = terminated or truncated
    episodes.append(episode)

# copy the connector pipeline for PPO

example_batch = AddOneTsToEpisodesAndTruncate(env.observation_space, env.action_space)(rl_module=rl_module, batch={}, episodes=episodes, shared_data={})
example_batch = AddObservationsFromEpisodesToBatch(env.observation_space, env.action_space, as_learner_connector=True)(rl_module=rl_module, batch=example_batch, episodes=episodes, explore=True, shared_data={})
example_batch = AddColumnsFromEpisodesToTrainBatch(env.observation_space, env.action_space)(rl_module=rl_module, batch=example_batch, episodes=episodes, explore=True, shared_data={})
example_batch = AddTimeDimToBatchAndZeroPad(env.observation_space, env.action_space)(rl_module=rl_module, batch=example_batch, episodes=episodes, explore=True, shared_data={})
example_batch = AddStatesFromEpisodesToBatch(env.observation_space, env.action_space)(rl_module=rl_module, batch=example_batch, episodes=episodes, explore=True, shared_data={})
example_batch = BatchIndividualItems(env.observation_space, env.action_space)(rl_module=rl_module, batch=example_batch, episodes=episodes, explore=True, shared_data={})
example_batch = NumpyToTensor(env.observation_space, env.action_space, as_learner_connector=True)(rl_module=rl_module, batch=example_batch, episodes=episodes, explore=True, shared_data={})

multi_rl_module = MultiRLModule(observation_space=env.observation_space, action_space=env.action_space)
multi_rl_module.add_module(module_id="tristan_test", module=rl_module)

example_batch = GeneralAdvantageEstimation(env.observation_space, env.action_space, gamma=0.99, lambda_=0.95)(rl_module=multi_rl_module, batch={"tristan_test": example_batch}, episodes=episodes, explore=True, shared_data={})
```

To illustrate the first issue, `torch.mean(example_batch["tristan_test"]["advantages"])` is (essentially) 0 but the value you'd get from eg possibly_masked_mean in the PPO loss (`torch.sum(example_batch["tristan_test"]["advantages"] * example_batch["tristan_test"]["loss_mask"]) / torch.sum(example_batch["tristan_test"]["loss_mask"])`) is not.

To illustrate the second issue (point 2 in my subsidiary questions, which I am less sure about due to my relative inexperience with the internals of rllib), `example_batch["tristan_test"]["value_targets"][-1]` is 0 as expected but I do not find a 0 at the place corresponding to the boundary between the 1st and 2nd episode that I collected.

### Issue Severity

None

### PR DESCRIPTION
**Pull request by lonexreb**
## Why are these changes needed?

Closes #57989.

The reporter pointed out that ``GeneralAdvantageEstimation`` standardizes advantages over the entire unpadded batch, including the phantom bootstrap timesteps that ``AddOneTsToEpisodesAndTruncate`` appends and then explicitly masks out of the loss via ``Columns.LOSS_MASK = False``.

The connector documents the contract directly (see ``add_one_ts_to_episodes_and_truncate.py:151``):

```python
loss_mask = [True for _ in range(len_)] + [False]
```

The phantom step has an arbitrary advantage that does not correspond to a real transition. Including it in the mean/std biases the standardization in proportion to the ratio of bootstrap steps to real steps. The effect is worst for short episode chunks and BPTT-truncated recurrent rollouts, where bootstrap steps are a large fraction of the unpadded batch.

## What this PR does

Restricts the standardization stats to ``LOSS_MASK == True`` rows via numpy's ``mean(where=...)`` / ``std(where=...)`` API:

```python
if Columns.LOSS_MASK in batch[module_id]:
    loss_mask = unpad_data_if_necessary(
        episode_lens,
        convert_to_numpy(batch[module_id][Columns.LOSS_MASK]),
    ).astype(bool)
    mean = module_advantages.mean(where=loss_mask)
    std = module_advantages.std(where=loss_mask)
else:
    mean = module_advantages.mean()
    std = module_advantages.std()
module_advantages = (module_advantages - mean) / max(1e-4, std)
```

Shape of ``module_advantages`` is preserved; only the normalization constants change. The fallback path covers callers that swap ``AddOneTsToEpisodesAndTruncate`` out of the pipeline.

The original issue reporter sketched the same fix and noted that the pre-ConnectorV2 stack handled masking correctly; this restores parity.

## Related issue number

Closes #57989

## Checks

- [x] DCO sign-off
- [x] ``ruff check`` and ``black==22.10.0`` (Ray's pinned version) pass on the modified file

## Test plan

- [ ] CI passes (``buildkite/premerge``, ``buildkite/microcheck``, ``buildkite/release``)
- [ ] Existing PPO learning tests (CartPole sanity check etc.) keep converging — they were the baseline regression for the unmasked code path and should still converge with the masked stats (which are strictly more correct on bootstrap-padded batches)

### CODE REVIEW

File: rllib/connectors/learner/general_advantage_estimation.py [hunk 1]
```diff
@@ -144,10 +144,22 @@ def __call__(
...
+                loss_mask = unpad_data_if_necessary(
+                    episode_lens,
+                    convert_to_numpy(batch[module_id][Columns.LOSS_MASK]),
+                ).astype(bool)
+                mean = module_advantages.mean(where=loss_mask)
+                std = module_advantages.std(where=loss_mask)
+            else:
+                mean = module_advantages.mean()
+                std = module_advantages.std()
+            module_advantages = (module_advantages - mean) / max(1e-4, std)
```
[the remark this answers is not in the document]
Comment (lonexreb) [reply rllib/connectors/learner/general_advantage_estimation.py:163]: Applied in deb8afa — switched to `module_advantages[loss_mask].mean()` / `.std()` per your suggestion. Cleaner, version-portable, and avoids the per-element `where=` mask check during the reduction.

**Comment by pseudo-rnd-thoughts** [comment]
Fixed in https://github.com/ray-project/ray/pull/59958
